### PR TITLE
refactor: Membership list model for groups and landscapes

### PR DESCRIPTION
**Pull request by josebui**
## Description

This is a really big PR, but I organized the commits by section (for most of the changes) so ideally if you check commit by commit the changes isolated to specific components will be there so that you can chek the related changes, hopefully that is good enough to review this.

### Checklist
- [x] Corresponding issue has been opened
- [x] New tests added

### Related Issues
- https://github.com/techmatters/terraso-product/issues/415

### Related PRs
- https://github.com/techmatters/terraso-client-shared/pull/153
- https://github.com/techmatters/terraso-web-client/pull/1317

### CODE REVIEW

File: terraso_backend/apps/graphql/schema/groups_memberships.py [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+        group_slug = graphene.String(required=True)
+        membership_status = graphene.String()
+
+    @classmethod
+    def mutate_and_get_payload(cls, root, info, **kwargs):
+        user = info.context.user
+
+        user_role = (
+            kwargs["user_role"] if "user_role" in kwargs else group_collaboration_roles.ROLE_MEMBER
+        )
```
Comment (paulschreiber) [inline terraso_backend/apps/graphql/schema/groups_memberships.py:55]:
```suggestion
        user_role = kwargs.get("user_role", group_collaboration_roles.ROLE_MEMBER)
```

File: terraso_backend/apps/graphql/schema/story_maps_memberships.py [hunk 2]
```diff
@@ -0,0 +1,326 @@
...
+            logger.exception("Failure to verify JWT token", extra={"token": invite_token})
+            raise GraphQLNotAllowedException(
+                model_name=StoryMap.__name__, operation=MutationTypes.UPDATE
+            )
+
+        try:
+            membership = Membership.objects.get(id=decoded_token["membershipId"])
+        except Exception as error:
+            logger.error(
+                "Attempt to approve Membership, but it was not found",
```
Comment (paulschreiber) [inline terraso_backend/apps/graphql/schema/story_maps_memberships.py:134]:
```suggestion
                "Attempt to approve Story Map Membership, but it was not found",
```

File: terraso_backend/apps/graphql/schema/story_maps_memberships.py [hunk 3]
```diff
@@ -0,0 +1,326 @@
...
+                extra={"membership_id": decoded_token["membershipId"], "error": error},
+            )
+            raise GraphQLNotFoundException(model_name=Membership.__name__)
+
+        if not user and membership.pending_email is None:
+            logger.error(
+                "Attempt to approve a Membership, but user was not found",
```
Comment (paulschreiber) [inline terraso_backend/apps/graphql/schema/story_maps_memberships.py:141]:
```suggestion
                "Attempt to approve a Story Map Membership, but user was not found",
```

File: terraso_backend/apps/graphql/schema/story_maps_memberships.py [hunk 4]
```diff
@@ -0,0 +1,326 @@
...
+        if not rules.test_rule(
+            "allowed_to_approve_story_map_membership_with_token",
+            request_user,
+            {
+                "decoded_token": decoded_token,
+                "membership": membership,
+            },
+        ):
+            logger.info(
+                "Attempt to approve a Membership, but user has no permission",
```
Comment (paulschreiber) [inline terraso_backend/apps/graphql/schema/story_maps_memberships.py:165]:
```suggestion
                "Attempt to approve a Story Map Membership, but user has no permission",
```

File: terraso_backend/apps/graphql/schema/story_maps_memberships.py [hunk 5]
```diff
@@ -0,0 +1,326 @@
...
+                ),
+            )
+
+        try:
+            membership.membership_list.approve_membership(
+                membership_id=membership.id,
+            )
+        except Exception as error:
+            logger.error(
+                "Attempt to approve Membership, but there was an error",
```
Comment (paulschreiber) [inline terraso_backend/apps/graphql/schema/story_maps_memberships.py:189]:
```suggestion
                "Attempt to approve Story Map Membership, but there was an error",
```

File: terraso_backend/apps/graphql/schema/story_maps_memberships.py [hunk 6]
```diff
@@ -0,0 +1,326 @@
...
+        membership_id = graphene.String(required=True)
+
+    @classmethod
+    def mutate_and_get_payload(cls, root, info, **kwargs):
+        membership_id = kwargs["membership_id"]
+        user = info.context.user
+
+        if not user:
+            logger.error(
+                "Attempt to approve a Membership, but user was not found",
```
Comment (paulschreiber) [inline terraso_backend/apps/graphql/schema/story_maps_memberships.py:212]:
```suggestion
                "Attempt to approve a Story Map Membership, but user was not found",
```

File: terraso_backend/apps/graphql/schema/story_maps_memberships.py [hunk 7]
```diff
@@ -0,0 +1,326 @@
...
+                extra=kwargs,
+            )
+            raise GraphQLNotFoundException(model_name=User.__name__)
+
+        try:
+            membership = Membership.objects.get(id=membership_id)
+        except Exception as error:
+            logger.error(
+                "Attempt to approve Membership, but it was not found",
```
Comment (paulschreiber) [inline terraso_backend/apps/graphql/schema/story_maps_memberships.py:221]:
```suggestion
                "Attempt to approve Story Map Membership, but it was not found",
```

File: terraso_backend/apps/graphql/schema/story_maps_memberships.py [hunk 8]
```diff
@@ -0,0 +1,326 @@
...
+                extra={"membership_id": membership_id, "error": error},
+            )
+            raise GraphQLNotFoundException(model_name=Membership.__name__)
+
+        if membership.user != user:
+            logger.error(
+                "Attempt to approve Membership, but user does not match",
```
Comment (paulschreiber) [inline terraso_backend/apps/graphql/schema/story_maps_memberships.py:228]:
```suggestion
                "Attempt to approve Story Map Membership, but user does not match",
```
Under what circumstances does the user not match?

File: terraso_backend/apps/graphql/schema/story_maps_memberships.py [hunk 9]
```diff
@@ -0,0 +1,326 @@
...
+        if not rules.test_rule(
+            "allowed_to_approve_story_map_membership",
+            user,
+            {
+                "story_map": story_map,
+                "membership": membership,
+            },
+        ):
+            logger.info(
+                "Attempt to approve a Membership, but user has no permission",
```
Comment (paulschreiber) [inline terraso_backend/apps/graphql/schema/story_maps_memberships.py:254]:
```suggestion
                "Attempt to approve a Story Map Membership, but user has no permission",
```

File: terraso_backend/apps/graphql/schema/story_maps_memberships.py [hunk 10]
```diff
@@ -0,0 +1,326 @@
...
+                model_name=Membership.__name__, operation=MutationTypes.UPDATE
+            )
+
+        try:
+            membership.membership_list.approve_membership(
+                membership_id=membership.id,
+            )
+        except Exception as error:
+            logger.error(
+                "Attempt to approve Membership, but there was an error",
```
Comment (paulschreiber) [inline terraso_backend/apps/graphql/schema/story_maps_memberships.py:267]:
```suggestion
                "Attempt to approve Story Membership, but there was an error",
```

File: terraso_backend/apps/graphql/schema/story_maps_memberships.py [hunk 11]
```diff
@@ -0,0 +1,326 @@
...
+                "Attempt to delete Story Map Memberships, but story map was not found",
+                extra={"story_map_id": story_map_id, "story_map_slug": story_map_slug},
+            )
+            raise GraphQLNotFoundException(model_name=StoryMap.__name__)
+
+        try:
+            membership = story_map.membership_list.memberships.get(id=membership_id)
+        except Membership.DoesNotExist:
+            logger.error(
+                "Attempt to delete Story Map Memberships, but it was not found",
```
Comment (paulschreiber) [inline terraso_backend/apps/graphql/schema/story_maps_memberships.py:305]:
```suggestion
                "Attempt to delete Story Map Membership, but it was not found",
```

File: terraso_backend/apps/graphql/schema/story_maps_memberships.py [hunk 12]
```diff
@@ -0,0 +1,326 @@
...
+        user = info.context.user
+        membership_id = kwargs["id"]
+        story_map_id = kwargs["story_map_id"]
+        story_map_slug = kwargs["story_map_slug"]
+
+        try:
+            story_map = StoryMap.objects.get(slug=story_map_slug, story_map_id=story_map_id)
+        except StoryMap.DoesNotExist:
+            logger.error(
+                "Attempt to delete Story Map Memberships, but story map was not found",
```
Comment (paulschreiber) [inline terraso_backend/apps/graphql/schema/story_maps_memberships.py:296]:
```suggestion
                "Attempt to delete Story Map Membership, but story map was not found",
```

File: terraso_backend/apps/graphql/schema/story_maps_memberships.py [hunk 13]
```diff
@@ -0,0 +1,326 @@
...
+        if not rules.test_rule(
+            "allowed_to_delete_story_map_membership",
+            user,
+            {
+                "story_map": story_map,
+                "membership": membership,
+            },
+        ):
+            logger.info(
+                "Attempt to delete Story Map Memberships, but user has no permission",
```
Comment (paulschreiber) [inline terraso_backend/apps/graphql/schema/story_maps_memberships.py:319]:
```suggestion
                "Attempt to delete Story Map Membership, but user has no permission",
```

File: terraso_backend/apps/graphql/schema/story_maps_memberships.py [hunk 14]
```diff
@@ -0,0 +1,326 @@
...
+        cls.validate_role(kwargs["user_role"], [ROLE_EDITOR])
+
+        story_map_id = kwargs["story_map_id"]
+        story_map_slug = kwargs["story_map_slug"]
+
+        try:
+            story_map = StoryMap.objects.get(slug=story_map_slug, story_map_id=story_map_id)
+        except Exception as error:
+            logger.error(
+                "Attempt to save Story Map Memberships, but story map was not found",
```
Comment (paulschreiber) [inline terraso_backend/apps/graphql/schema/story_maps_memberships.py:66]:
```suggestion
                "Attempt to save Story Map Membership, but story map was not found",
```

**Review by paulschreiber — CHANGES_REQUESTED** [review]

File: terraso_backend/apps/graphql/schema/groups_memberships.py [hunk 15]
```diff
@@ -0,0 +1,183 @@
...
+
+        cls.validate_role(user_role, group_collaboration_roles.ALL_ROLES)
+
+        group_slug = kwargs["group_slug"]
+
+        try:
+            group = Group.objects.get(slug=group_slug)
+        except Exception as error:
+            logger.error(
+                "Attempt to save Story Map Memberships, but story map was not found",
```
Comment (paulschreiber) [inline terraso_backend/apps/graphql/schema/groups_memberships.py:65]:
```suggestion
                "Attempt to save Group Membership, but story map was not found",
```

File: terraso_backend/apps/graphql/schema/groups_memberships.py [hunk 16]
```diff
@@ -0,0 +1,183 @@
...
+        if not rules.test_rule(
+            "allowed_group_managers_count",
+            user,
+            {
+                "group": group,
+                "membership": membership,
+            },
+        ):
+            logger.info(
+                "Attempt to update a Membership, but cannot remove last manager",
```
Comment (paulschreiber) [inline terraso_backend/apps/graphql/schema/groups_memberships.py:173]:
```suggestion
                "Attempt to update a Group Membership, but cannot remove last manager",
```

File: terraso_backend/apps/graphql/schema/groups_memberships.py [hunk 17]
```diff
@@ -0,0 +1,183 @@
...
+        if not rules.test_rule(
+            "allowed_to_delete_group_membership",
+            user,
+            {
+                "group": group,
+                "membership": membership,
+            },
+        ):
+            logger.info(
+                "Attempt to delete Group Memberships, but user lacks permission",
```
Comment (paulschreiber) [inline terraso_backend/apps/graphql/schema/groups_memberships.py:157]:
```suggestion
                "Attempt to delete Group Membership, but user lacks permission",
```

File: terraso_backend/apps/graphql/schema/landscapes_memberships.py [hunk 18]
```diff
@@ -0,0 +1,154 @@
...
+
+        cls.validate_role(kwargs["user_role"], landscape_collaboration_roles.ALL_ROLES)
+
+        landscape_slug = kwargs["landscape_slug"]
+
+        try:
+            landscape = Landscape.objects.get(slug=landscape_slug)
+        except Exception as error:
+            logger.error(
+                "Attempt to save Landscape Memberships, but landscape was not found",
```
Comment (paulschreiber) [inline terraso_backend/apps/graphql/schema/landscapes_memberships.py:60]:
```suggestion
                "Attempt to save Landscape Membership, but landscape was not found",
```

File: terraso_backend/apps/graphql/schema/landscapes_memberships.py [hunk 19]
```diff
@@ -0,0 +1,154 @@
...
+        if not rules.test_rule(
+            "allowed_to_delete_landscape_membership",
+            user,
+            {
+                "landscape": landscape,
+                "membership": membership,
+            },
+        ):
+            logger.info(
+                "Attempt to delete Landscape Memberships, but user lacks permission",
```
Comment (paulschreiber) [inline terraso_backend/apps/graphql/schema/landscapes_memberships.py:128]:
```suggestion
                "Attempt to delete Landscape Membership, but user lacks permission",
```

**Review by paulschreiber — CHANGES_REQUESTED** [review]

File: terraso_backend/apps/core/permission_rules.py [hunk 20]
```diff
@@ -60,42 +71,29 @@ def allowed_to_delete_landscape_group(user, landscape_group):
     )
 
 
-@rules.predicate
-def allowed_to_change_membership(user, membership_group_id):
-    return user.is_group_manager(membership_group_id)
+def validate_managers_count(user, membership, entity):
```
[the remark this answers is not in the document]
Comment (josebui) [reply terraso_backend/apps/core/permission_rules.py:74]: @paulschreiber yes, that is it

File: terraso_backend/apps/graphql/schema/groups.py [hunk 21]
```diff
@@ -122,18 +103,26 @@ class Input:
...
     def mutate_and_get_payload(cls, root, info, **kwargs):
         user = info.context.user
 
         if not cls.is_update(kwargs):
             kwargs["created_by"] = user
 
-        if "membership_type" in kwargs:
-            kwargs["membership_type"] = Group.get_membership_type_from_text(
-                kwargs["membership_type"]
+        membership_type = kwargs.pop("membership_type") if "membership_type" in kwargs else None
```
[the remark this answers is not in the document]
Comment (josebui) [reply terraso_backend/apps/graphql/schema/groups.py:113]: @paulschreiber this is not the same `pop` removes the key from the dictionary `get` is not doing that

[the remark this answers is not in the document]
Comment (paulschreiber) [reply terraso_backend/apps/graphql/schema/groups.py:113]: Ah. Looks like `pop()` also supports a default value.

**Review by paulschreiber — APPROVED** [review]

**Pull request merged**